core/templates/services/html-length.service.ts

```diff
@@ -50,51 +65,129 @@ export class HtmlLengthService {
     }
     const sanitizedHtml = this.sanitizer.sanitize(
       SecurityContext.HTML, htmlString) as string;
-    let totalWeight = this.calculateBaselineLength(sanitizedHtml);
+    // Identify custom tags using regex on the original HTML string.
+    const customTags = htmlString.match(CUSTOM_TAG_REGEX);
+    let customTagsLength = (customTags) ? customTags.length : 0;
+
+    /* CustomTagsLength is being passed to calculateBaselineLength()
+    method to ensure try-catch block should not raise error if
+    we use non-interactive RTE */
+    let totalWords = this.calculateBaselineLength(
+      sanitizedHtml, CALCULATION_TYPE_WORD, customTagsLength);
+
+    if (customTags) {
+      for (const customTag of customTags) {
+        totalWords += this.getWeightForNonTextNodes(
+          customTag, CALCULATION_TYPE_WORD);
+      }
+    }
+    return totalWords;
+  }
+
+  computeHtmlLengthInCharacters(htmlString: string): number {
+    if (!htmlString) {
+      this.loggerService.error('Empty string was passed to compute length');
+      return 0;
+    }
 
+    const sanitizedHtml = this.sanitizer.sanitize(
+      SecurityContext.HTML, htmlString) as string;
     // Identify custom tags using regex on the original HTML string.
-    const customTagRegex = /<oppia-noninteractive-(?:math|image)[^>]*>/g;
-    const customTags = htmlString.match(customTagRegex);
+    const customTags = htmlString.match(CUSTOM_TAG_REGEX);
+    let customTagsLength = (customTags) ? customTags.length : 0;
+
+    /* CustomTagsLength is being passed to calculateBaselineLength()
+    method to ensure try-catch block should not raise error if
+    we use non-interactive RTE */
+    let totalCharacters = this.calculateBaselineLength(
+      sanitizedHtml, CALCULATION_TYPE_CHARACTER, customTagsLength);
 
     if (customTags) {
       for (const customTag of customTags) {
-        totalWeight += this.getWeightForNonTextNodes(customTag);
+        totalCharacters += this.getWeightForNonTextNodes(
+          customTag, CALCULATION_TYPE_CHARACTER);
       }
     }
-    return totalWeight;
+    return totalCharacters;
   }
 
-  private calculateBaselineLength(sanitizedHtml: string): number {
+  calculateBaselineLength(
+      sanitizedHtml: string,
+      calculationType: CalculationType, CustomTagsLength: number): number {
     let domparser = new DOMParser();
-    let dom = domparser.parseFromString(sanitizedHtml, 'text/html');
+    let dom: Document;
+    try {
+      dom = domparser.parseFromString(sanitizedHtml, 'text/html');
+      if (dom.body.children.length === 0 && !CustomTagsLength) {
+        throw new Error(
+          'No HTML tags found. Ensure ' +
+          'that a valid string that includes HTML tags is provided.');
+      }
+    } catch (error) {
+      throw new Error(
+        'Failed to parse HTML string.' +
+        ' Ensure that a valid string that includes HTML tags is provided.');
+    }
     let totalWeight = 0;
```

Comment:
@seanlip To simplify we can remove ``getWeightForTextNodes()``, as its just there for text extraction which can be done in ``calculateBaselineLength()``, if we remove loop node, then it won't check for kind of tags, In present it won't make much difference removing that loop node but future developer might get stuck if any other tags is introduced in text nodes.

Reply:
Why would the developer get stuck? I don't understand the scenario here that you're concerned about, could you please elaborate?

(In general, simpler code is easier to maintain, so that's why I need to dig into this. We don't want this to become unnecessarily complex code that devs have trouble following.)

Reply:
I have removed it to make it simpler; my point was if need to remove or add any text node, then they might get stuck, but if we will introduce anything new in the future, then at that time we can put conditions accordingly. Thanks!